platforms/allwinner-d1/d1-core/src/drivers/smhc.rs

```diff
@@ -0,0 +1,754 @@
+        let prg_clk = || {
+            smhc.smhc_cmd.write(|w| {
+                w.wait_pre_over().wait();
+                w.prg_clk().change();
+                w.cmd_load().set_bit();
+                w
+            });
+
+            while smhc.smhc_cmd.read().cmd_load().bit_is_set() {
+                core::hint::spin_loop()
+            }
+
+            smhc.smhc_rintsts.write(|w| unsafe { w.bits(0xFFFFFFFF) });
+        };
+
+        // Reset the SD/MMC controller
+        smhc.smhc_ctrl.modify(|_, w| w.soft_rst().reset());
+        while smhc.smhc_ctrl.read().soft_rst().is_reset() {
+            core::hint::spin_loop();
+        }
+
+        // Reset the FIFO
+        smhc.smhc_ctrl.modify(|_, w| w.fifo_rst().reset());
+        while smhc.smhc_ctrl.read().fifo_rst().is_reset() {
+            core::hint::spin_loop();
+        }
+
+        // Global interrupt disable
+        smhc.smhc_ctrl.modify(|_, w| w.ine_enb().disable());
+
+        prg_clk();
+
+        // Set card clock = module clock / 2
+        smhc.smhc_clkdiv.modify(|_, w| w.cclk_div().variant(1));
+
+        // Set the sample delay to 0 (as done in Linux and Allwinner BSP)
+        smhc.smhc_smap_dl.write(|w| {
+            w.samp_dl_sw().variant(0);
+            w.samp_dl_sw_en().set_bit();
+            w
+        });
+
+        // Enable the card clock
+        smhc.smhc_clkdiv.modify(|_, w| w.cclk_enb().on());
+
+        prg_clk();
+
+        // Default bus width after power up or idle is 1-bit
+        smhc.smhc_ctype.write(|w| w.card_wid().b1());
+        // Blocksize is fixed to 512 bytes
+        smhc.smhc_blksiz.write(|w| unsafe { w.bits(512) });
+
+        Self {
+            smhc,
+            isr: &SMHC0_ISR,
+            int: (int, isr),
+        }
+    }
+
+    /// # Safety
+    /// The descriptor chain needs to live at least as long as the DMA transfer.
+    /// Additionally, their content (e.g., the validity of the buffers they point to)
+    /// also needs to be verified by the user.
+    unsafe fn prepare_dma(&self, descriptor: &idmac::Descriptor, byte_cnt: u32) {
+        self.smhc.smhc_ctrl.modify(|_, w| {
+            w.dma_enb().set_bit();
+            w.dma_rst().set_bit();
+            w
+        });
+        while self.smhc.smhc_ctrl.read().dma_rst().bit_is_set() {}
+
+        // Configure the address of the first DMA descriptor
+        // Right-shift by 2 because it is a *word-address*.
+        self.smhc
+            .smhc_dlba
+            .write(|w| unsafe { w.bits((descriptor as *const _ as u32) >> 2) });
+
+        // Set number of bytes that will be read or written.
+        self.smhc.smhc_bytcnt.write(|w| unsafe { w.bits(byte_cnt) });
+
+        // Soft reset of DMA controller
+        self.smhc.smhc_idmac.write(|w| w.idmac_rst().set_bit());
+
+        // Configure the burst size and TX/RX trigger level
+        self.smhc.smhc_fifoth.write(|w| {
+            w.tx_tl().variant(8);
+            w.rx_tl().variant(7);
```

Comment:
minor nit: can we add what burst size and trigger levels were selecting to the comment? it's not immediately clear what "8" and "7" are saying here — is that the actual value of the trigger level or is it selecting something else?